aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu

```diff
@@ -106,6 +106,74 @@ void triangular_solve_batched_cublas(Tensor& A, Tensor& B, Tensor& infos, bool u
   });
 }
 
+template <typename scalar_t>
+inline void apply_gels_batched(const Tensor& A, Tensor& B, Tensor& infos) {
+// AMD ROCm backend is implemented via rewriting all CUDA calls to HIP
+// rocBLAS does not implement BLAS-like extensions of cuBLAS, they're in rocSOLVER
+// rocSOLVER is currently not used in ATen, therefore we raise an error in this case
+#ifndef CUDART_VERSION
+  TORCH_CHECK(false, "torch.linalg.lstsq: Batched version is supported only with cuBLAS backend.")
+#else
+  auto trans = CUBLAS_OP_N;
+  auto m = cuda_int_cast(A.size(-2), "m");
+  auto n = cuda_int_cast(A.size(-1), "n");
+
+  auto nrhs = cuda_int_cast(B.size(-1), "nrhs");
+  // cuBLAS from cuda10 and older doesn't work with nrhs == 0 (cuda11 works)
+  // so we need to put this early return
+  if (nrhs == 0) {
+    return;
+  }
+
+  auto batch_size = cuda_int_cast(batchCount(B), "batch_size");
+  auto lda = std::max<int>(1, m);
+  auto ldb = std::max<int>(1, m);
+
+  // cuBLAS's requirement
+  TORCH_CHECK(
+    m >= n,
+    "torch.linalg.lstsq: only overdetermined systems (input.size(-2) >= input.size(-1)) are allowed on CUDA with cuBLAS backend.");
+
+  // cuBLAS documentation says:
+  // Matrices Aarray[i] should not overlap; otherwise, undefined behavior is expected.
+  // explicitly broadcast the batch dimensions of A
+  IntArrayRef A_batch_sizes(A.sizes().data(), A.dim() - 2);
+  IntArrayRef B_batch_sizes(B.sizes().data(), B.dim() - 2);
+  std::vector<int64_t> expand_batch_portion = at::infer_size(A_batch_sizes, B_batch_sizes);
+  expand_batch_portion.insert(expand_batch_portion.end(), {A.size(-2), A.size(-1)});
+  Tensor A_expanded = A.expand({expand_batch_portion});
+  Tensor A_broadcasted = cloneBatchedColumnMajor(A_expanded);
```

Comment:
I'm a bit concerned with this broadcasting behavior. (which is a really nice feature!) Do we have a documentation entry mentioning about the linear algebra broadcasting (for batch dimension)? Also, can you please check if we have a test for this broadcasting?

I think linalg broadcasting sounds like a big feature that we should either enable for most operators, or just don't mention in the doc. 😅

Reply:
There is a test `test_linalg_lstsq_batch_broadcasting` in `test/test_linalg.py`
https://github.com/pytorch/pytorch/blob/f84f2063b43a32fbe88c7e4de7e1019a00adb302/test/test_linalg.py#L250

Broadcasting is applicable only for solve functions because they are the only ones that take two interacting inputs.
I don't think that it is documented separately somewhere, it's just expected that this function is broadcastable as many others in PyTorch are. lstsq still doesn't implement broadcasting if some dimensions do not exist (general behavior described here https://pytorch.org/docs/stable/notes/broadcasting.html). Other solve functions materialize explicitly the broadcasted inputs as done here for cuBLAS, and lstsq is the first function that tries to address the memory inefficiencies.

We have two issues open for discussing the broadcasting of linear algebra solvers https://github.com/pytorch/pytorch/issues/52915 and https://github.com/pytorch/pytorch/issues/49252.

Reply:
Thanks. That's nice!